tyrian/js/src/main/scala/tyrian/runtime/TyrianRuntime.scala

```diff
@@ -139,36 +138,51 @@ final class TyrianRuntime[F[_]: Async, Model, Msg](
   def toVNode(html: Html[Msg]): VNode =
     html match
       case Tag(name, attrs, children) =>
-        val as = js.Dictionary(
+        val as: List[(String, String)] =
           attrs.collect {
             case Attribute(n, v)   => (n, v)
             case NamedAttribute(n) => (n, "")
-          }: _*
-        )
-
-        val props =
-          js.Dictionary(attrs.collect { case Property(n, v) => (n, v) }: _*)
+          }
 
-        val events =
-          js.Dictionary(attrs.collect { case Event(n, msg) =>
-            (n, fun((e: dom.Event) => onMsg(msg.asInstanceOf[dom.Event => Msg](e))))
-          }: _*)
+        val props: List[(String, PropValue)] =
+          attrs.collect { case Property(n, v) => (n, v) }
 
-        val childrenElem: List[VNodeParam] =
-          children.map {
-            case t: Text            => VNodeParam.Text(t.value)
-            case subHtml: Html[Msg] => VNodeParam.Node(toVNode(subHtml))
+        val events: List[(String, dom.Event => Unit)] =
+          attrs.collect { case Event(n, msg) =>
+            (n, (e: dom.Event) => onMsg(msg.asInstanceOf[dom.Event => Msg](e)))
           }
 
-        h(name, obj(props = props, attrs = as, on = events))(childrenElem: _*)
+        val data: VNodeData =
+          VNodeData.builder
+            .withProps(props: _*)
+            .withAttrs(as: _*)
+            .withOn(events: _*)
+            .build
```

Comment:
This I don't like. I can't provide nice collections, and I have to use this funky builder thing. I want to just do:
```scala
VNodeData.empty.withProps(props)
```
Where props is a `List`|`Array`|`js.Array`|`Whatever`.

Reply:
I think the thing here is: Someone mad enough to use this lib directly would benefit from the set up you have here. If you're building a library on top of it though, it's just inefficient. Though to be fair if efficiency is my concern, I just probably just use the main constructor or do a single `VnodeData.empty` and... set the `vars`?

Reply:
I agree with you on this one. At the very least it's worth offering the alternate constructor.

> Though to be fair if efficiency is my concern, I just probably just use the main constructor or do a single `VnodeData.empty` and... set the `vars`?

I'd be shocked if the Scala.js optimizer doesn't inline this to the same thing. Actually, we can make this explicit by adding `@inline` annotations. It's a mutable builder so it's just friendly syntax for setting the `vars` on a `VNodeData.empty`.